src/Transaction.cc

```diff
@@ -0,0 +1,293 @@
+/* Copyright (c) 2015 Stanford University
+ *
+ * Permission to use, copy, modify, and distribute this software for any
+ * purpose with or without fee is hereby granted, provided that the above
+ * copyright notice and this permission notice appear in all copies.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR(S) DISCLAIM ALL WARRANTIES
+ * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL AUTHORS BE LIABLE FOR
+ * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ */
+
+#include "ClientTransactionTask.h"
+#include "Transaction.h"
+
+namespace RAMCloud {
+
+/**
+ * Constructor for a transaction.
+ *
+ * \param ramcloud
+ *      Overall information about the calling client.
+ */
+Transaction::Transaction(RamCloud* ramcloud)
+    : ramcloud(ramcloud)
+    , taskPtr(new ClientTransactionTask(ramcloud))
+    , commitStarted(false)
+{
+}
+
+/**
+ * Commits the transaction defined by the operations performed on this
+ * transaction (read, remove, write).  This method blocks until a decision is
+ * reached but not until the decisions are synced.
+ *
+ * \return
+ *      True if the transaction was able to commit.  False otherwise.
+ */
+bool
+Transaction::commit()
+{
+    if (!commitStarted) {
+        commitStarted = true;
+        ramcloud->transactionManager.addTransactionTask(taskPtr);
+    }
+
+    ClientTransactionTask* task = taskPtr.get();
+
+    while (task->getDecision() == WireFormat::TxDecision::INVALID) {
+        ramcloud->poll();
+    }
+
+    if (task->getStatus() != STATUS_OK) {
+        ClientException::throwException(HERE, task->getStatus());
+    }
+
+    return (task->getDecision() == WireFormat::TxDecision::COMMIT);
+}
+
+/**
+ * Block until the decision of this transaction commit is accepted by all
+ * participant servers.  If the commit has not yet occurred and a decision is
+ * not yet reached, this method will also start the commit.
+ */
+void
+Transaction::sync()
```

Comment:
Why is this method exposed, rather than just being part of syncAndCommit?  I'm not sure why a client would invoke this method. Let's discuss.